cpp/include/ucxx/remote_key.h

```diff
@@ -0,0 +1,212 @@
+   * // remoteKey is `std::shared_ptr<ucxx::RemoteKey>`
+   * auto serializedRemoteKey = remoteKey->serialize();
+   * @endcode
+   *
+   * @throws std::runtime_error if checksum of the serialized object fails.
+   *
+   * @returns The serialized remote key.
+   */
+  void deserialize(const SerializedRemoteKey& serializedHeader);
+
+ public:
+  /**
+   * @brief Constructor for `std::shared_ptr<ucxx::RemoteKey>` from local memory handle.
+   *
+   * The constructor for a `std::shared_ptr<ucxx::RemoteKey>` object from a local
+   * `std::shared_ptr<ucxx::MemoryHandle>`, mapping a local memory buffer to be made
+   * accessible from a remote endpoint to perform RMA (Remote Memory Access) on the memory.
+   *
+   * @code{.cpp}
+   * // `memoryHandle` is `std::shared_ptr<ucxx::MemoryHandle>`
+   * auto remoteKey = memoryHandle->createRemoteKey();
+   *
+   * // Equivalent to line above
+   * // auto remoteKey = ucxx::createRemoteKeyFromMemoryHandle(memoryHandle);
+   * @endcode
+   *
+   * @throws ucxx::Error if `ucp_rkey_pack` fails.
+   *
+   * @param[in] memoryHandle the memory handle mapped on the local process.
+   *
+   * @returns The `shared_ptr<ucxx::RemoteKey>` object
+   */
+  friend std::shared_ptr<RemoteKey> createRemoteKeyFromMemoryHandle(
+    std::shared_ptr<MemoryHandle> memoryHandle);
+
+  /**
+   * @brief Constructor for `std::shared_ptr<ucxx::RemoteKey>` from remote.
+   *
+   * The constructor for a `std::shared_ptr<ucxx::RemoteKey>` object from a serialized
+   * `std::shared_ptr<ucxx::RemoteKey>`, mapping a remote memory buffer to be made
+   * accessible via a local endpoint to perform RMA (Remote Memory Access) on the memory.
+   *
+   * @code{.cpp}
+   * // `serializedRemoteKey` is `ucxx::SerializedRemoteKey>`, created on a remote worker
+   * // after a call to `ucxx::RemoteKey::serialize()` and transferred over-the-wire.
+   * auto remoteKey = ucxx::createRemoteKeyFromSerialized(serializedRemoteKey);
+   *
+   * // Equivalent to line above
+   * // auto remoteKey = ucxx::createRemoteKeyFromMemoryHandle(memoryHandle);
+   * @endcode
+   *
+   * @throws ucxx::Error if `ucp_ep_rkey_unpack` fails.
+   *
+   * @param[in] endpoint            the `std::shared_ptr<Endpoint>` parent component.
+   * @param[in] serializedRemoteKey the remote key that was serialized by the owner of
+   *                                the memory handle and transferred over-the-wire for
+   *                                reconstruction and remote access.
+   *
+   * @returns The `shared_ptr<ucxx::RemoteKey>` object
+   */
+  friend std::shared_ptr<RemoteKey> createRemoteKeyFromSerialized(
+    std::shared_ptr<Endpoint> endpoint, SerializedRemoteKey serializedRemoteKey);
+
+  ~RemoteKey();
+
+  /**
+   * @brief Get the underlying `ucp_rkey_h` handle.
+   *
+   * Lifetime of the `ucp_rkey_h` handle is managed by the `ucxx::RemoteKey` object and
+   * its ownership is non-transferrable. Once the `ucxx::RemoteKey` is destroyed the address
```

Comment:
*question*: Does this suggest the `RemoteKey` should not be a `Component` (and hence we shouldn't hand out `shared_ptr`s to it)? Or do I misunderstand?

Reply:
Ah, thanks.